File: 01-arrays-and-hashing/contains-duplicate/containsDuplicate.ts
/**
 * Contains Duplicate
 * ===================================================================
 * Read README.md first. Return true if any value in `nums` appears at
 * least twice, else false.
 *
 * TODO(you): Restate the problem in ONE sentence:
 *   -
 *
 * TODO(you): Approach — which structure answers "seen this before?" in O(1)?
 *   -  (hint: TypeScript's built-in `Set`)
 *
 * --- Fill these in AFTER you solve it -----------------------------
 * Time complexity:  O(?)   //
 * Space complexity: O(?)   //
 */

import assert from "node:assert";

export function containsDuplicate(nums: number[]): boolean {
  // -------------------------------------------------------------------
  // TODO(you): Step 1 — track the values you've already seen.
  // TODO(you): Step 2 — as you scan, if you meet one you've seen, return true.
  // TODO(you): Step 3 — if you finish the scan, there were no duplicates.
  // -------------------------------------------------------------------
  return false; // ← replace with your solution
}

// ---- Self-check tests (do NOT edit) ---------------------------------
// Run:  npx tsx containsDuplicate.ts
assert.strictEqual(containsDuplicate([1, 2, 3, 1]), true);
assert.strictEqual(containsDuplicate([1, 2, 3, 4]), false);
assert.strictEqual(containsDuplicate([1, 1, 1, 3, 3, 4, 3, 2, 4, 2]), true);
assert.strictEqual(containsDuplicate([7]), false);
assert.strictEqual(containsDuplicate([-1, -2, -3, -1]), true);
console.log("✅ All tests passed!");

File: 01-arrays-and-hashing/contains-duplicate/contains_duplicate.py
"""
Contains Duplicate
=====================================================================
Read README.md first. Return True if any value in `nums` appears at
least twice, else False.

TODO(you): Restate the problem in ONE sentence:
    -

TODO(you): Approach — which data structure answers "seen this before?"
           in O(1)?
    -

--- Fill these in AFTER you solve it -------------------------------
Time complexity:  O(?)   #
Space complexity: O(?)   #
"""
from typing import List


def contains_duplicate(nums: List[int]) -> bool:
    # -----------------------------------------------------------------
    # TODO(you): Step 1 — track the values you've already seen.
    see_num = []
    # TODO(you): Step 2 — as you scan, if you meet one you've seen, return True.
    for num in nums:
        if num in see_num:
            return True
        see_num.append(num)
    # TODO(you): Step 3 — if you finish the scan, there were no duplicates.
    return False    
    # -----------------------------------------------------------------

# ---- Self-check tests (do NOT edit) --------------------------------
# Run:  python3 contains_duplicate.py
if __name__ == "__main__":
    assert contains_duplicate([1, 2, 3, 1]) is True
    assert contains_duplicate([1, 2, 3, 4]) is False
    assert contains_duplicate([1, 1, 1, 3, 3, 4, 3, 2, 4, 2]) is True
    assert contains_duplicate([7]) is False
    assert contains_duplicate([-1, -2, -3, -1]) is True
    print("✅ All tests passed!")
